sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala

```diff
@@ -165,19 +165,19 @@ class HiveClientSuite(version: String, allVersions: Seq[String])
     assert(client.getDatabase("temporary").properties.contains("flag"))
 
     // test alter database location
+    val oldDatabasePath = database.locationUri.getPath
     val tempDatabasePath2 = Utils.createTempDir().toURI
-    // Hive support altering database location since HIVE-8472.
+    client.alterDatabase(database.copy(locationUri = tempDatabasePath2))
+    val uriInCatalog = client.getDatabase("temporary").locationUri
+    assert("file" === uriInCatalog.getScheme)
+
     if (version == "3.0" || version == "3.1") {
-      client.alterDatabase(database.copy(locationUri = tempDatabasePath2))
-      val uriInCatalog = client.getDatabase("temporary").locationUri
-      assert("file" === uriInCatalog.getScheme)
+      // Hive support altering database location since HIVE-8472
       assert(new Path(tempDatabasePath2.getPath).toUri.getPath === uriInCatalog.getPath,
         "Failed to alter database location")
     } else {
-      val e = intercept[AnalysisException] {
-        client.alterDatabase(database.copy(locationUri = tempDatabasePath2))
-      }
-      assert(e.getMessage.contains("does not support altering database location"))
+      // .. otherwise, the command should be non-effective against older versions of Hive
+      assert(oldDatabasePath === uriInCatalog.getPath, "Expected database location to be unchanged")
```

Comment:
If there is no exception from HMS side in this case, we had better give a warning from Spark side here.

Reply:
Sorry for misleading you, @sunchao . I was wondering this case which didn't throw exception. If there is no sign of failures (or no-op), the customer will surprise at the next session.

Reply:
Can we do double-check the location inside this method after invoking `shim.alterDatabase`?

https://github.com/apache/spark/blob/52e2717c2d1b6e1f449de5714b6e202074bac26f/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L357-L366

Reply:
Good idea! Instead of warning, I'm thinking maybe we should throw exception when the location is not changed.

Reply:
+1 for throwing exception!

Reply:
Updated the PR